pkg/pillar/cmd/volumemgr/create.go

```diff
@@ -33,55 +34,82 @@ func createVdiskVolume(ctx *volumemgrContext, status types.VolumeStatus,
 
 	created := false
 
+	persistFsType := ctx.persistType
+
 	// this is the target location, where we expect the volume to be
 	filelocation := status.PathName()
-	if _, err := os.Stat(filelocation); err == nil {
-		errStr := fmt.Sprintf("Can not create %s for %s: exists",
-			filelocation, status.Key())
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-
-	// use the edge-containers library to extract the data we need
-	puller := registry.Puller{
-		Image: ref,
-	}
-
-	casClient, err := cas.NewCAS(casClientType)
-	if err != nil {
-		err = fmt.Errorf("Run: exception while initializing CAS client: %s", err.Error())
-		return created, "", err
-	}
-	defer casClient.CloseClient()
-	ctrdCtx, done := casClient.CtrNewUserServicesCtx()
-	defer done()
 
-	resolver, err := casClient.Resolver(ctrdCtx)
-	if err != nil {
-		errStr := fmt.Sprintf("error getting CAS resolver: %v", err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-
-	// create a writer for the file where we want
-	f, err := os.Create(filelocation)
-	if err != nil {
-		errStr := fmt.Sprintf("error creating target file at %s: %v", filelocation, err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-	defer f.Close()
-
-	if _, _, err := puller.Pull(&registry.FilesTarget{Root: f, AcceptHash: true}, 0, false, os.Stderr, resolver); err != nil {
-		errStr := fmt.Sprintf("error pulling %s from containerd: %v", ref, err)
-		log.Error(errStr)
-		return created, "", errors.New(errStr)
-	}
-
-	// Do we need to expand disk?
-	if err := maybeResizeDisk(filelocation, status.MaxVolSize); err != nil {
-		log.Error(err)
-		return created, "", err
+	if persistFsType != types.PersistZFS {
+		if _, err := os.Stat(filelocation); err == nil {
+			errStr := fmt.Sprintf("Can not create %s for %s: exists",
+				filelocation, status.Key())
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
+	}
+
+	switch persistFsType {
+	case types.PersistZFS:
+		pathToFile, _, err := getVolumeFilePathAndVSize(ctx, status)
+		if err != nil {
+			errStr := fmt.Sprintf("Error obtaining file for zvol at volume %s, error=%v",
+				status.Key(), err)
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
+		zVolName := status.ZVolName(types.VolumeZFSPool)
+		zVolDevice := zfs.GetZVolDeviceByDataset(zVolName)
+		if zVolDevice == "" {
+			errStr := fmt.Sprintf("Error finding zfs zvol %s", zVolName)
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
+		if err := diskmetrics.ConvertImg(log, pathToFile, zVolDevice, "raw"); err != nil {
+			errStr := fmt.Sprintf("Error converting %s to zfs zvol %s: %v",
+				pathToFile, zVolDevice, err)
+			log.Error(errStr)
+			return created, "", errors.New(errStr)
+		}
+		filelocation = zVolDevice
```

Comment:
Shouldn't you set the qouta to status.MaxVolSize somewhere here?

Reply:
Quota in this case is the size of zvol we create during preparation step.

Reply:
But I didn't see that looking at MaxVolSize. Did I miss it?

Reply:
It is here https://github.com/lf-edge/eve/pull/2134/files#diff-b08d0791d3bbddce21f342a820c0df0989aab8958de3176a21c0505d270a25c4R65-R71

Reply:
A function called get() shouldn't modify things. 
I'd suggest a function which is visible here so that it is clear that for both Ext* and zfs we take MaxVolSize into account.

Reply:
Hmm, but current implementation of `getVolumeFilePathAndVSize` returns vSize. The problem here: in case of qcow2 we can have MaxVolSize equals with provided qcow2 file size, which one may be compressed. In case of converting qcow2->raw(zvol) we will have an error here, So, we forced to check virtualsize here and create zvol with size equals the virtualsize of provided qcow2 image.

Reply:
See my suggestions on how to refactor this.